### PR TITLE
aya: Add PinnedProgram

### PR DESCRIPTION
**Pull request by dave-tucker**
This commit adds PinnedProgram which allows the creation of a Program from a path on bpffs. This is useful to be able to call `attach` or other APIs for programs that are already loaded to the kernel. Unfortunately figuring out the concrete type of program is hard given the information in the kernel so this may not work for all program types.

Signed-off-by: Dave Tucker <dave@dtucker.co.uk>

### CODE REVIEW

File: aya/src/programs/mod.rs [hunk 1]
```diff
@@ -395,10 +400,209 @@ impl Drop for Program {
...
+            }
+        };
+        Ok(PinnedProgram {
+            path: PathBuf::from(path.as_ref()),
+            inner: p,
+        })
+    }
+
+    /// Removes the pinned program from bpffs.
+    pub fn unpin(&self) -> Result<(), io::Error> {
```
Comment (dave-tucker) [inline aya/src/programs/mod.rs:584]: @alessandrod I really want to have this as:
```rust
pub fn unpin(self) -> Result<Program, io::Error> {
    fs::remove_file(self.path.clone())?;
    Ok(self.inner)
}
```

What stopped me was that something is calling `Program::drop` :thinking: 
Even wrapping the inner `Program` in an `Option<Program>` and calling `take()` on it doesn't seem to help.

File: test/integration-test/src/tests/load.rs [hunk 2]
```diff
@@ -157,11 +167,48 @@ fn pin_lifecycle() -> anyhow::Result<()> {
...
+    }
+
+    // should still be loaded since prog was pinned
+    assert_loaded!("pass");
+
+    // 2. Load program from bpffs
+    {
+        let mut pinned = PinnedProgram::from_pin("/sys/fs/bpf/aya-xdp-test-prog")?;
+
+        // I can perform an attach operatoin here
```
Comment (vadorovsky) [inline test/integration-test/src/tests/load.rs:200]:
```suggestion
        // I can perform an attach operation here
```

File: aya/src/obj/btf/btf.rs [hunk 3]
```diff
@@ -399,6 +399,18 @@ impl Btf {
                     fixed_ty.name_offset = 0;
                     types.types[i] = BtfType::Ptr(fixed_ty)
                 }
+                // Fixup STRUCT for Rust
```
Comment (vadorovsky) [inline aya/src/obj/btf/btf.rs:402]: We should probably create issues for those fixups, to not lose track of them and think of proper LLVM/rustc fixes. I'm curious - what are the examples of such names?

**Review by vadorovsky — COMMENTED** [review]
The code looks good. I just have one question/remark and one nit.

**Comment by dave-tucker** [comment]
Closing in favour of #444 since I accidentally included extra commits in this branch that bpfd is dependent on.

**Pull request closed without merging**